File: Code/Day05.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AoC.Code
{
    class Day05
    {
        public static int Solve(IEnumerable<int> indata)
        {
            var a = indata.ToArray();
            var i = 0;
            var steps = 0;
            while (i >= 0 && i < a.Length)
            {
                steps++;
                var jump = a[i];
                a[i]++;
                i += jump;
            }
            return steps;
        }


        public static int Solve2(IEnumerable<int> indata)
        {
            var a = indata.ToArray();
            var i = 0;
            var steps = 0;
            while (i >= 0 && i < a.Length)
            {
                steps++;
                var jump = a[i];
                if(jump >= 3)
                    a[i]--;
                else
                    a[i]++;
                i += jump;
            }
            return steps;
        }
    }
}

File: AoC.FSharp/Day05.fs
namespace AoC.FSharp

module Day05 =
    let solve (x: seq<int>) = 
        let ary = Seq.toArray x
        let max = ary.Length
        let mutable pos = 0
        let mutable counter = 0

        while pos >= 0 && pos < max do
            let jump = ary.[pos]
            Array.set ary pos (jump + 1)
            pos <- pos + jump
            counter <- counter + 1

        counter

    let solve2 (x: seq<int>) = 
        let ary = Seq.toArray x
        let max = ary.Length
        let mutable pos = 0
        let mutable counter = 0

        while pos >= 0 && pos < max do
            let jump = ary.[pos]
            Array.set ary pos (if jump >=3 then jump - 1 else jump + 1)
            pos <- pos + jump
            counter <- counter + 1

        counter